crates/sui-framework/sources/governance/staking_pool.move

```diff
@@ -0,0 +1,253 @@
+        delegation_token_supply: Supply<DelegationToken>,
+        // TODO: add fields for implementing bonding periods if necessary
+    }
+
+    /// An inactive staking pool associated with an inactive validator.
+    /// Only withdraws can be made from this pool.
+    struct InactiveStakingPool has key {
+        id: UID,
+        pool: StakingPool,
+    }
+
+    /// The staking pool token.
+    struct DelegationToken has drop {}
+
+    /// A self-custodial delegation object, serving as evidence that the delegator
+    /// has delegated to a staking pool.
+    struct Delegation has key {
+        id: UID,
+        /// The sui address of the validator associated with the staking pool this object delgates to.
+        validator_address: address,
+        /// The epoch at which the staking pool started operating.
+        pool_starting_epoch: u64,
+        /// The ealiest epoch at which this delegation can be activated.
+        earliest_activation_epoch: u64,
+        /// The staked SUI tokens.
+        principal: Balance<SUI>,
+        /// If the stake comes from a Coin<SUI>, this field is None. If it comes from a LockedCoin<SUI>, this
+        /// field will record the original lock expiration epoch, to be used when unstaking.
+        sui_token_lock: Option<EpochTimeLock>,
+        /// The pool tokens representing the amount of rewards the delegator can get back when they withdraw
+        /// from the pool. If this field is `none`, that means the delegation hasn't been activated yet.
+        pool_tokens: Option<Balance<DelegationToken>>,
+    }
+
+    /// Create a new, empty staking pool.
+    public(friend) fun new(validator_address: address, starting_epoch: u64) : StakingPool {
+        StakingPool {
+            validator_address,
+            starting_epoch,
+            sui_balance: 0,
+            rewards_pool: balance::zero(),
+            delegation_token_supply: balance::create_supply(DelegationToken {}),
+        }
+    }
+
+    /// Add rewards (in SUI) to the staking pool. 
+    /// Notice here we add SUI tokens but do not mint new pool tokens. Thus the exchange rate between SUI and pool tokens goes up.
+    public(friend) fun add_rewards(pool: &mut StakingPool, rewards: Balance<SUI>) {
+        pool.sui_balance = pool.sui_balance + balance::value(&rewards);
+        balance::join(&mut pool.rewards_pool, rewards);
+        // TODO: implement potential bonding period related bookkeeping
+    }
+
+    /// Request to delegate to a staking pool. The delegation doesn't get activated and the delegator doesn't get
+    /// the pool tokens until they activate the delegation on or after the `earliest_activation_epoch`.
+    public(friend) fun request_add_delegation(pool: &mut StakingPool, stake: Balance<SUI>, sui_token_lock: Option<EpochTimeLock>, ctx: &mut TxContext) {
+        assert!(balance::value(&stake) > 0, 0);
+        let delegation = Delegation {
```

Comment:
Is it possible for user to have 2 Delegations for the same validator?

Reply:
Yes. Or more than 2.